File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int widthOfBinaryTree(TreeNode* root) {
         queue<pair<TreeNode *, long long>> q;
        if (root == NULL)
            return 0;
        q.push({root, 0});
        int ans = 0;
        while (!q.empty())
        {
            auto f = q.front();
            int mmin = f.second;
            auto n = q.size();
            int last, first;
            for (int i = 0; i < n; i++)
            {
                auto cur = q.front();
                q.pop();
                auto adj = cur.second - mmin;
                auto node = cur.first;
                if (i == 0)
                {
                    first = adj;
                }
                last =adj;
                if (node->left)
                {
                    q.push({node->left, 2 * adj + 1});
                }
                if (node->right)
                {
                    q.push({node->right, 2 * adj + 2});
                }
            }
            ans = max(last - first + 1, ans);
        }
        return ans;
    }
};

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        q = deque()
        q.append((root,1))
        depth =  0
        while q:
            # print(q[-1][1],q[0][1])
            ans = max(ans,q[-1][1]-q[0][1]+1)
            lenQ = len(q)
            begin = q[0][1]
            for i in range(lenQ):
                cur,idx = q.popleft()
                idx-=begin
                if cur.left:
                    q.append((cur.left,idx*2))
                if cur.right:
                    q.append((cur.right,1+(idx*2)))
            depth+=1
        return ans
        
        